---
title: "Pre-processing Functional HUMAnN All samples from CPM table - Yogurt study"
output: 
  html_notebook:
    number_sections: yes
    toc: yes
    toc_depth: 3
---



```{r message=FALSE}
library(phyloseq)
library(dplyr)
library(readxl)
library(tidyr)
library(ggplot2)
library(mia)
library(microbiome)
library(patchwork)
library(ggExtra)

dir0 <- "Q:/IPE-A-Epidemiologie/Virginie_Stanislas/Projects/"
dir1 <- "Q:/IPE-P-Joghurtstudie/Joghurt und Haferflocken/Analysis/"
dir <- paste(dir1, "DataPreparation/Microbiome_DATA_Kun_Huang/", sep="")

source(file = paste(dir1, "Utils.R", sep=""))
source(file = paste(dir0, "R functions/Utils_metagenomics.R", sep=""))
source(file = paste(dir0, "R functions/Utils_longitudinal.R", sep=""))
source(file = paste(dir0, "R functions/Utils_crossOver.R", sep=""))
source(file = paste(dir0, "R functions/code_utils.R", sep=""))
```



# Loading tables
## Metadata
```{r}
metadata <- read_excel(paste(dir, "Metadata.xlsx", sep=""), sheet=1)
metadata <- as.data.frame(metadata)

metadata$Sample_ID <- gsub("-", "", metadata$Sample_ID)
metadata$Treatment <- case_when(metadata$Timepoint %in% c("T1", "T3") ~ "Before",
                                metadata$Timepoint %in% c("T2", "T4") ~ "After",
                                TRUE ~ NA_character_)
metadata$Intervention <- case_when(metadata$Timepoint %in% c("T1", "T2") & metadata$Gruppe == "A" ~ "Yogurt",
                                   metadata$Timepoint %in% c("T3", "T4") & metadata$Gruppe == "A" ~ "YogurtOatmeal",
                                   metadata$Timepoint %in% c("T1", "T2") & metadata$Gruppe == "B" ~ "YogurtOatmeal",
                                   metadata$Timepoint %in% c("T3", "T4") & metadata$Gruppe == "B" ~ "Yogurt",
                                TRUE ~ NA_character_)

metadata$Inter_treat <- paste(metadata$Treatment, metadata$Intervention, sep=".")
metadata$Inter_treat <- factor(metadata$Inter_treat, levels = c("Before.Yogurt", "After.Yogurt", "Before.YogurtOatmeal", "After.YogurtOatmeal"), ordered=T)
metadata$Treatment <- factor(metadata$Treatment, levels = c("Before", "After"), ordered=T)
metadata$Timepoint <- factor(metadata$Timepoint, levels = c("T1", "T2", "T3", "T4"), ordered=T)
metadata$Gruppe <- as.factor(metadata$Gruppe)
metadata$Participant_ID <- as.factor(metadata$Participant_ID)
metadata$Geschlecht <- as.factor(metadata$Geschlecht)
metadata$Intervention <- as.factor(metadata$Intervention)

rownames(metadata) <- metadata[, "Sample_ID"]
```

```{r}
dim(metadata) # 440   11
```


## Pathway and description

```{r}
load("JH_191601genes_CPM.Rdata")
dim(genes_JH)
```

```{r}
# rearrange column, row names
colnames(genes_JH) <- gsub("_Abundance.CPM", "", colnames(genes_JH))
colnames(genes_JH) <- gsub("Ext_86_", "", colnames(genes_JH)) # one sample with names: Ext_86_JH277_Abundance.CPM
rownames(genes_JH) <- genes_JH$Genes
genes_JH$Total_ab <- rowSums(genes_JH[, -c(1:2)], na.rm=T)
genes_JH <- genes_JH[order(genes_JH$Total_ab, decreasing=T),]


# separate pathway description to another file
desc_df <- genes_JH[, c("Genes", "Descr", "Total_ab")] 

genes_JH$Genes <- NULL
genes_JH$Descr <- NULL
genes_JH$Total_ab <- NULL
```


```{r}
# we didn't detected NA 
sum(is.na(genes_JH)) # 0
```

```{r}
# replace NA by 0
# genes_JH[is.na(genes_JH)] <- 0
```

```{r}
summary(colSums(genes_JH))
```

```{r}
hist(colSums(genes_JH))
```


```{r}
# Normalization
genes_JH_CPM <- sweep(genes_JH, 2, colSums(genes_JH), FUN="/")* 1000000
```


```{r}
dim(genes_JH)
```

```{r}
summary(colSums(genes_JH_CPM))
```


```{r}
hist(colSums(genes_JH_CPM))
```

```{r}
inters_names <- intersect(colnames(genes_JH_CPM), rownames(metadata))
cat("Samples present in HUMAnN and absent from Metadata: ")
colnames(genes_JH_CPM)[!(colnames(genes_JH_CPM) %in% inters_names)]

cat("Samples present in Metadata and absent from HUMAnN: ")
rownames(metadata)[!(rownames(metadata) %in% inters_names)]
```



# Creating objects
## Tree Summarized Experiment

### 946461 genes renorm
```{r}
metadata <- metadata[order(metadata$Sample_ID), ]
tse_genes_renorm <- TreeSummarizedExperiment(assays =  SimpleList(CPM=as.matrix(genes_JH_CPM)),
                                        colData = DataFrame(metadata),
                                        rowData = DataFrame(desc_df))
```

```{r}
tse_genes_renorm
```


### 946461 genes
```{r}
metadata <- metadata[order(metadata$Sample_ID), ]
tse_genes <- TreeSummarizedExperiment(assays =  SimpleList(CPM=as.matrix(genes_JH)),
                                        colData = DataFrame(metadata),
                                        rowData = DataFrame(desc_df))
```

```{r}
tse_genes
```



## Phyloseq

### 946461 genes renorm
```{r}
OTU = otu_table(as.matrix(genes_JH_CPM), taxa_are_rows = TRUE)
TAX = tax_table(desc_df[, 1:2])
rownames(TAX) <- TAX[,"ta1"]
colnames(TAX) <- c("Genes", "Descr")
SAM = sample_data(metadata)
```

```{r}
dim(OTU)
```


```{r}
inters_names <- intersect(colnames(OTU), rownames(metadata))
cat("Samples present in Metaphlan and absent from Metadata: ")
colnames(OTU)[!(colnames(OTU) %in% inters_names)]

cat("Samples present in Metadata and absent from Metaphlan: ")
rownames(metadata)[!(rownames(metadata) %in% inters_names)]
```


```{r}
physeq_genes_renorm  = phyloseq(OTU, TAX, SAM)
physeq_genes_renorm  
```


```{r}
# Print sample / taxa lost when creating phyloseq object
if(!ncol(OTU) == ncol(otu_table(physeq_genes_renorm))){
  inters_names <- intersect(colnames(OTU), colnames(otu_table(physeq_genes_renorm)))
  cat("Phyloseq sample lost: ")
  colnames(OTU)[!(colnames(OTU) %in% inters_names)] %>% print()
}
if(!nrow(OTU) == nrow(otu_table(physeq_genes_renorm))){
  inters_names <- intersect(rownames(OTU), rownames(otu_table(physeq_genes_renorm)))
  cat("Phyloseq taxa lost: ")
  rownames(OTU)[!(rownames(OTU) %in% inters_names)] %>% print()
}
```


### 946461 genes
```{r}
OTU = otu_table(as.matrix(genes_JH), taxa_are_rows = TRUE)
TAX = tax_table(desc_df[, 1:2])
rownames(TAX) <- TAX[,"ta1"]
colnames(TAX) <- c("Genes", "Descr")
SAM = sample_data(metadata)
```

```{r}
dim(OTU)
```


```{r}
inters_names <- intersect(colnames(OTU), rownames(metadata))
cat("Samples present in Metaphlan and absent from Metadata: ")
colnames(OTU)[!(colnames(OTU) %in% inters_names)]

cat("Samples present in Metadata and absent from Metaphlan: ")
rownames(metadata)[!(rownames(metadata) %in% inters_names)]
```


```{r}
physeq_genes = phyloseq(OTU, TAX, SAM)
physeq_genes 
```


```{r}
# Print sample / taxa lost when creating phyloseq object
if(!ncol(OTU) == ncol(otu_table(physeq_genes))){
  inters_names <- intersect(colnames(OTU), colnames(otu_table(physeq_genes)))
  cat("Phyloseq sample lost: ")
  colnames(OTU)[!(colnames(OTU) %in% inters_names)] %>% print()
}
if(!nrow(OTU) == nrow(otu_table(physeq_genes))){
  inters_names <- intersect(rownames(OTU), rownames(otu_table(physeq_genes)))
  cat("Phyloseq taxa lost: ")
  rownames(OTU)[!(rownames(OTU) %in% inters_names)] %>% print()
}
```

# Prevalence and abundance

## 946461 genes renorm
```{r}
total_abundance <- rowSums(genes_JH_CPM) 
max_abundance <- apply(genes_JH_CPM, 1, max)
preval <- getPrevalence(tse_genes_renorm, assay_name="CPM")

df <- data.frame(Species= names(total_abundance), total_abundance=total_abundance, max_abundance=max_abundance, preval=preval)
```


```{r fig.height=7, fig.width=10}
p1 <- ggplot(df, aes(x=total_abundance)) + geom_histogram(bins=30, color="white") + scale_x_log10()
p2 <- ggplot(df, aes(x=max_abundance)) + geom_histogram(bins=30, color="white") + scale_x_log10()
p3 <- ggplot(df, aes(x=preval)) + geom_histogram(bins=30, color="white") + scale_x_log10()
wrap_plots(p1, p2, p3, ncol=2)
```

## 946461 genes 
```{r}
total_abundance <- rowSums(genes_JH) 
max_abundance <- apply(genes_JH, 1, max)
preval <- getPrevalence(tse_genes, assay_name="CPM")

df <- data.frame(Species= names(total_abundance), total_abundance=total_abundance, max_abundance=max_abundance, preval=preval)
```


```{r fig.height=7, fig.width=10}
p1 <- ggplot(df, aes(x=total_abundance)) + geom_histogram(bins=30, color="white") + scale_x_log10()
p2 <- ggplot(df, aes(x=max_abundance)) + geom_histogram(bins=30, color="white") + scale_x_log10()
p3 <- ggplot(df, aes(x=preval)) + geom_histogram(bins=30, color="white") + scale_x_log10()
wrap_plots(p1, p2, p3, ncol=2)
```



# Normalization 

## Check total number of reads

```{r fig.height=4, fig.width=7, message=FALSE}
sum <-  colSums(genes_JH)
mean <- colMeans(genes_JH)
sam <- data.frame(sum=sum, mean=mean)
pp <- ggplot(sam, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "blue")+  ggtitle("Distribution of sequencing depth") 
pp


sum <-  colSums(genes_JH_CPM)
mean <- colMeans(genes_JH_CPM)
sam <- data.frame(sum=sum, mean=mean)
pp <- ggplot(sam, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "blue")+  ggtitle("Distribution of sequencing depth") 
pp
```


# Saving objects

```{r}
save(tse_genes, physeq_genes, file=paste(dir, "Functional/TSE_Physeq_191601genes_fromCPM.Rdata", sep=""))
```

```{r}
save(tse_genes_renorm, physeq_genes_renorm, file=paste(dir, "Functional/TSE_Physeq_191601genes_fromCPM_renorm.Rdata", sep=""))
```



