---
title: "Analysis of functional pathways"
output: 
  html_notebook:
    number_sections: yes
    toc: yes
    toc_depth: 4
---


<style>
.main-container { width: 1200px; max-width:2800px;}
</style>





# Loading

```{r message=FALSE}
library(phyloseq)
library(ggplot2)
library(plyr)
library(ape)
library(knitr)
library(vegan)
library(tidyr)
library(gridExtra)
# library(lemon)
library(RColorBrewer)
library(dplyr)
library(cowplot)
library(scater)
library(ggnewscale)
library(mia)
library(miaViz)
library(kableExtra)
library(patchwork)
library(microViz)
library(lme4)
library(emmeans)
library(flextable)
library(ggpubr)
library(reshape)
library(tictoc)
```


```{r message=FALSE}
dir0 <- "Q:/IPE-A-Epidemiologie/Virginie_Stanislas/Projects/"
dir <- "Q:/IPE-P-Joghurtstudie/Joghurt und Haferflocken/Analysis/"
path_analysis <- paste(dir, "DataAnalysis/Results_Complete_Analysis/ResultsFunctional/On_pathways_AllSamples/", sep="")

source(file = paste(dir, "Utils.R", sep=""))
source(file = paste(dir0, "R functions/Utils_metagenomics.R", sep=""))
source(file = paste(dir0, "R functions/Utils_longitudinal.R", sep=""))
source(file = paste(dir0, "R functions/Utils_crossOver.R", sep=""))
source(file = paste(dir0, "R functions/code_utils.R", sep=""))
```


```{r}
### Load data with already ordination (saved at the end of ordination section)
### contains tse, tse_pathOnly
load(paste(path_analysis, "tse_allwithordination.Rdata", sep=""))
```


```{r message=FALSE}
load(paste(dir, "DataPreparation/Microbiome_DATA_Kun_Huang/Functional/TSE_Physeq_Pathway.Rdata", sep=""))

tse0 <- tse_pathway  # 606 440
physeq0 <- physeq_pathway # 606 440

# Remove 2 samples from ID016 not correctly prepared by the participant
tse <- tse0[, c(!(tse0$Sample_ID  %in% c("JH379", "JH380")))]
physeq <- ps_filter(physeq0, !(Sample_ID %in% c("JH379", "JH380"))) 


tse_pathOnly <- tse[!(rowData(tse)$Pathway  %in% c("UNMAPPED", "UNINTEGRATED")),] # 604 438
physeq_pathOnly <- subset_taxa(physeq, !(Pathway  %in% c("UNMAPPED", "UNINTEGRATED"))) # 603 438 
# 1 pathway less in phyloseq because PWY-5109 is not in any samples anymore after fileting out the two samples from ID016
# phyloseq removed it but not tse
```


```{r message=FALSE}
# removed ordered factor to allow some plotting functions to work properly
physeq_noFact <-  physeq
sample_data(physeq_noFact)$Inter_treat <- as.character(sample_data(physeq_noFact)$Inter_treat )
sample_data(physeq_noFact)$Inter_treat <- as.factor(sample_data(physeq_noFact)$Inter_treat )

physeq_pathOnly_noFact <-  physeq_pathOnly
sample_data(physeq_pathOnly_noFact)$Inter_treat <- as.character(sample_data(physeq_pathOnly_noFact)$Inter_treat )
sample_data(physeq_pathOnly_noFact)$Inter_treat <- as.factor(sample_data(physeq_pathOnly_noFact)$Inter_treat )
```


```{r}
# proband_colors <- brewer.pal(n = length(unique(sample_data(mydata.psq)$Proband)), name = "Set1")
# names(proband_colors) <- unique(sample_data(mydata.psq)$Proband)

cols <- c("#017b74", "#73ffec", "#ad0053", "#ffcacf", "#ff8d3f", "#ffdf91", "#00c14e","#abffc2", "#6042ed", "#90b9ff", "#700054","#ff8bfc", "#3d8b00", "#bcff7c", "#8b3700", "#ffba91", "#ff4e2a", "#000e26", "#71a800", "#577eff", "#02d0d3", "#d00039", "#dba500", "#ac72ff", "#e0c9ff", "#002e66", "#006dbd", "#02a0c8", "#db37ec",  "#e5ffd5", "#004a20", "#637500", "#ec00bf", "#434d00","#ff77a8", "#002f17", "#f1ff4b", "#533500", "#7ea2ff", "#000d4e", "#ff41a0", "#003636", "#271800", "#E41A1C", "#377EB8", "#4DAF4A", "#984EA3", "#FF7F00", "#A65628", "#F781BF", "#FFFF00FF", "#999999")

cols_Inter_treat <- c(After.Yogurt="#7570B3", After.YogurtOatmeal="#D95F02", Before.Yogurt="#AAA8C5", Before.YogurtOatmeal="#DBB394")

cols_Participant_Id <- rep(cols, 3)
names(cols_Participant_Id) <- unique(colData(tse)[, "Participant_ID"])
```


# Exploration

## Prevalence x Abundance

**prevalence**: frequency of samples where the taxa was detected

**abundance**: number of reads assigned to a taxa (on average or in total)

```{r}
rowData(tse)$prevalence <- getPrevalence(tse, detection = 0, sort = FALSE, assay.type = "CPM", as_relative = TRUE)
rowData(tse)$abundance_sum <- rowSums(assays(tse)[["CPM"]])
rowData(tse)$abundance_mean <- rowMeans(assays(tse)[["CPM"]])
rowData(tse)$abundance_max <- apply(assays(tse)[["CPM"]], 1, max)

rowData(tse_pathOnly)$prevalence <- getPrevalence(tse_pathOnly, detection = 0, sort = FALSE, assay.type = "CPM", as_relative = TRUE)
rowData(tse_pathOnly)$abundance_sum <- rowSums(assays(tse_pathOnly)[["CPM"]])
rowData(tse_pathOnly)$abundance_mean <- rowMeans(assays(tse_pathOnly)[["CPM"]])
rowData(tse_pathOnly)$abundance_max <- apply(assays(tse_pathOnly)[["CPM"]], 1, max)
```


```{r fig.height=7, fig.width=9}
## Prevalence x abundance on the complete species data
ggplot(as.data.frame(rowData(tse)), aes(abundance_sum, prevalence)) +
  geom_point(alpha = 0.5) +
  geom_rug(alpha = 0.1) +
  scale_x_continuous(labels = scales::label_number(), name = "Total Abundance") +
  scale_y_continuous(
    labels = scales::label_percent(), breaks = scales::breaks_pretty(n = 9),
    name = "Prevalence (%)",
    sec.axis = sec_axis(
      trans = ~ . * ncol(tse), breaks = scales::breaks_pretty(n = 9),
      name = "Prevalence (N samples)"
    )) + 
  ggrepel::geom_text_repel(
    data = function(df) filter(df, abundance_sum > 300 | prevalence > 0.6),
  mapping = aes(label = Pathway), size = 2.5, min.segment.length = 0)+
   theme_bw()

ggplot(as.data.frame(rowData(tse_pathOnly)), aes(abundance_sum, prevalence)) +
  geom_point(alpha = 0.5) +
  geom_rug(alpha = 0.1) +
  scale_x_continuous(labels = scales::label_number(), name = "Total Abundance") +
  scale_y_continuous(
    labels = scales::label_percent(), breaks = scales::breaks_pretty(n = 9),
    name = "Prevalence (%)",
    sec.axis = sec_axis(
      trans = ~ . * ncol(tse), breaks = scales::breaks_pretty(n = 9),
      name = "Prevalence (N samples)"
    )) + 
  ggrepel::geom_text_repel(
    data = function(df) filter(df, abundance_sum > 300 | prevalence > 0.6),
  mapping = aes(label = Pathway), size = 2.5, min.segment.length = 0)+
   theme_bw()
```



```{r fig.height=10, fig.width=12}
## Prevalence x abundance on the complete species data - axis log transformed
ggplot(as.data.frame(rowData(tse)), aes(abundance_max, prevalence)) +
  geom_point(alpha = 0.5) +
  geom_rug(alpha = 0.1) +
  scale_x_continuous(trans=scales::pseudo_log_trans(base = 10), labels = scales::label_number(), name = "Total Abundance") +
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
    labels = scales::label_percent(), breaks = scales::breaks_pretty(n = 9),
    name = "Prevalence (%)",
    sec.axis = sec_axis(
      trans = ~ . * ncol(tse), breaks = scales::breaks_pretty(n = 9),
      name = "Prevalence (N samples)"
    )) + 
  ggrepel::geom_text_repel(
  mapping = aes(label = Pathway), size = 2.5, min.segment.length = 0)+
   theme_bw() 

ggplot(as.data.frame(rowData(tse_pathOnly)), aes(abundance_max, prevalence)) +
  geom_point(alpha = 0.5) +
  geom_rug(alpha = 0.1) +
  scale_x_continuous(trans=scales::pseudo_log_trans(base = 10), labels = scales::label_number(), name = "Total Abundance") +
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
    labels = scales::label_percent(), breaks = scales::breaks_pretty(n = 9),
    name = "Prevalence (%)",
    sec.axis = sec_axis(
      trans = ~ . * ncol(tse), breaks = scales::breaks_pretty(n = 9),
      name = "Prevalence (N samples)"
    )) + 
  ggrepel::geom_text_repel(
  mapping = aes(label = Pathway), size = 2.5, min.segment.length = 0)+
   theme_bw() 
```

```{r paged.print=FALSE}
abundance_CPM <- rowSums(assays(tse)[["CPM"]])
abundance_CPM <- abundance_CPM[order(abundance_CPM, decreasing = T)]
head(abundance_CPM, 15)
```
```{r}
plot(abundance_CPM[-c(1:2)])
```

# Ordination



- **Distance/dissimilarity measures** 
  - **`r colorize("Bray-curtis", "#02a0c8")`**: take into account the abundance $BC_{ij}= 1 - \frac{2C_{ij}}{S_i + S_j}$ with $S_i$ : Total number of species in sample $i$, $S_j$ : Total number of species in sample $j$ and $C_{ij}$ : Sum of lesser value for those species common in both samples
  - **`r colorize("Unifrac", "#02a0c8")`**:  take into account the phylogenetic tree information
      - unweigthed unifrac: Measures the fraction of unique branch length / measures the relative length of those branches that lead exclusively to species present in only one of the two samples with respect to the total length of all branches in the tree.
    - weigthed unifrac: includes information on the relative abundance of each taxa
  - **`r colorize("Jaccard", "#02a0c8")`**: presence / absence, ignores abundance and phylogenetic information
  - **`r colorize("Aitchison", "#02a0c8")`**: take into consideration the compositional structure of the data. Dissimilarity measure calculated as the Euclidean distance between samples after performing a centered log ratio (clr) transformation.
  
- **Ordination/dimention reduction methods** 
  - **`r colorize("PCoA/MDS", "#02a0c8")`**: extension of Principal Components Analysis (PCA),  detect linear trends
  - **`r colorize("NMDS", "#02a0c8")`**: mon-metric multidimensional scaling can detect non-linear patterns
  - **`r colorize("TSNE", "#02a0c8")`**: 
  - **`r colorize("UMAP", "#02a0c8")`**: 
  
## On Species space - full data

```{r}
load(paste(path_analysis, "tse_allwithordination.Rdata", sep=""))
```


```{r}
### Strange plot when using tse, probably UNMAPPED and UNINTEGRATED explain all the variation on first axis??
tse <- runMDS(tse, FUN = vegan::vegdist, method = "bray", name = "PCoA_BC", exprs_values = "relabundance", keep_dist = T)
set.seed(40)

tse <- runNMDS(tse, FUN = vegan::vegdist, method = "bray", name = "NMDS_BC", exprs_values = "relabundance", keep_dist = T)
set.seed(40)


# Alternative methods to try
# not easy to change the metric for UMAP, no braycurtis dissimilarity in package uwot
set.seed(122)
tse <- runTSNE(tse, name = "TSNE", exprs_values="relabundance")
set.seed(122)
tse <- runUMAP(tse, name = "UMAP", exprs_values="relabundance")
```


```{r}
tse_pathOnly <- runMDS(tse_pathOnly, FUN = vegan::vegdist, method = "bray", name = "PCoA_BC", exprs_values = "relabundance", keep_dist = T)
set.seed(40)

tse_pathOnly <- runNMDS(tse_pathOnly, FUN = vegan::vegdist, method = "bray", name = "NMDS_BC", exprs_values = "relabundance", keep_dist = T)
set.seed(40)


# Alternative methods to try
# not easy to change the metric for UMAP, no braycurtis dissimilarity in package uwot
set.seed(122)
tse_pathOnly <- runTSNE(tse_pathOnly, name = "TSNE", exprs_values="relabundance")
set.seed(122)
tse_pathOnly <- runUMAP(tse_pathOnly, name = "UMAP", exprs_values="relabundance")
```

```{r}
reducedDim(tse, type ="PCoA_BC_pathOnly") <- reducedDim(tse_pathOnly, "PCoA_BC")
reducedDim(tse, type ="NMDS_BC_pathOnly") <- reducedDim(tse_pathOnly, "NMDS_BC")
reducedDim(tse, type ="TSNE_pathOnly") <- reducedDim(tse_pathOnly, "TSNE")
reducedDim(tse, type ="UMAP_pathOnly") <- reducedDim(tse_pathOnly, "UMAP")
```

```{r}
tse_for_ordination <- tse
add_name <- ""

tse_for_ordination <- tse_pathOnly
add_name <- "_pathOnly"
```


### PCoA Bray-Curtis

Almost the same representation when done on filtered data `tse_sub`.
```{r fig.height=12, fig.width=24, message=FALSE, warning=FALSE, paged.print=FALSE}
pp <- plot_ordination(tse_for_ordination, "PCoA_BC", sd_threshold=0.0875, data_type="Functional")
e <- attr(reducedDim(tse_for_ordination, "PCoA_BC"), "eig");
rel_eig <- e/sum(e[e>0])          
x_lab = paste("PCoA 1 (", round(100 * rel_eig[[1]],1), "%", ")", sep = "")
y_lab = paste("PCoA 2 (", round(100 * rel_eig[[2]],1), "%", ")", sep = "")

PartHsd_PCoA_BC <- pp$Participants_high_sd
p1 <- pp$p_gruppe + labs(x=x_lab, y=y_lab)
p2 <- pp$p_topdom
p3 <- pp$p_2ndtopdom
p4 <- pp$p_inter_treat + labs(x=x_lab, y=y_lab)
p5 <- pp$p_part + labs(x=x_lab, y=y_lab)
p6 <- pp$p_hv_part + labs(x=x_lab, y=y_lab)
```
Warning message: "'clr' includes negative values. Agglomeration of it might lead to meaningless values." comes from `plot_ordination()` when identifying the most dominant taxa for a specific rank. We can ignore it as we use 'relabudance' assays for identifying dominant taxa.

```{r fig.height=7, fig.width=10, message=FALSE, paged.print=FALSE}
p2
```

```{r}
pathway_multi <- names(abundance_CPM)[1:13]

p_list <- list()
for(i in seq(length(pathway_multi))){
pathway <- pathway_multi[i]
p_list[[i]] <- plot_ordination(tse, paste("PCoA_BC", add_name, sep=""), pathway=pathway, data_type="Functional")$p_1spec 
}
```


```{r}
ggsave(paste(path_analysis, "Plots/Ordination_plots/PCoA_BC", add_name,".png", sep=""), 
       p1 + p5 + p2+ p4 + p6 + p3+ plot_layout(guides = 'collect'), 
       width=30, height=15)


ggsave(paste(path_analysis, "Plots/Ordination_plots/PCoA_BC", add_name,"_withAbundance.png", sep=""), 
       wrap_plots(p_list, ncol=4) , 
       width=30, height=20)
```



### NMDS Bray-Curtis

Almost the same representation when done on filtered data `tse_sub`.
```{r fig.height=12, fig.width=24, message=FALSE, warning=FALSE, paged.print=FALSE}
pp <- plot_ordination(tse_for_ordination, "NMDS_BC", sd_threshold=0.05, data_type="Functional")
PartHsd_NMDS_BC <- pp$Participants_high_sd
p1 <- pp$p_gruppe 
p2 <- pp$p_topdom
p3 <- pp$p_2ndtopdom
p4 <- pp$p_inter_treat 
p5 <- pp$p_part 
p6 <- pp$p_hv_part
```


```{r fig.height=7, fig.width=10, message=FALSE, paged.print=FALSE}
p2
```

```{r}
pathway_multi <- names(abundance_CPM)[1:13]

p_list <- list()
for(i in seq(length(pathway_multi))){
pathway <- pathway_multi[i]
p_list[[i]] <- plot_ordination(tse, paste("NMDS_BC", add_name, sep=""), pathway=pathway, data_type="Functional")$p_1spec 
}
```


```{r}
ggsave(paste(path_analysis, "Plots/Ordination_plots/NMDS_BC", add_name,".png", sep=""), 
       p1 + p5 + p2+ p4 + p6 + p3+ plot_layout(guides = 'collect'), 
       width=30, height=15)

ggsave(paste(path_analysis, "Plots/Ordination_plots/NMDS_BC", add_name,"_withAbundance.png", sep=""), 
       wrap_plots(p_list, ncol=4) , 
       width=30, height=20)
```



### TSNE
```{r fig.height=12, fig.width=24, message=FALSE, warning=FALSE, paged.print=FALSE}
pp <- plot_ordination(tse_for_ordination, "TSNE", sd_threshold=10, data_type="Functional")
PartHsd_TSNE <- pp$Participants_high_sd
p1 <- pp$p_gruppe 
p2 <- pp$p_topdom
p3 <- pp$p_2ndtopdom
p4 <- pp$p_inter_treat 
p5 <- pp$p_part 
p6 <- pp$p_hv_part
```


```{r fig.height=7, fig.width=10, message=FALSE, paged.print=FALSE}
p2
```

```{r}
pathway_multi <- names(abundance_CPM)[1:13]

p_list <- list()
for(i in seq(length(pathway_multi))){
pathway <- pathway_multi[i]
p_list[[i]] <- plot_ordination(tse, paste("TSNE", add_name, sep=""), pathway=pathway, data_type="Functional")$p_1spec 
}
```


```{r}
ggsave(paste(path_analysis, "Plots/Ordination_plots/TSNE", add_name,".png", sep=""),
       p1 + p5 + p2+ p4 + p6 + p3+ plot_layout(guides = 'collect'), 
       width=30, height=15)

ggsave(paste(path_analysis, "Plots/Ordination_plots/TSNE", add_name,"_withAbundance.png", sep=""), 
       wrap_plots(p_list, ncol=4) , 
       width=30, height=20)
```


### UMAP



```{r fig.height=12, fig.width=24, message=FALSE, warning=FALSE}
pp <- plot_ordination(tse_for_ordination, "UMAP", sd_threshold=2, data_type="Functional")
PartHsd_UMAP <- pp$Participants_high_sd
p1 <- pp$p_gruppe
p2 <- pp$p_topdom
p3 <- pp$p_2ndtopdom
p4 <- pp$p_inter_treat 
p5 <- pp$p_part 
p6 <- pp$p_hv_part
```


```{r fig.height=7, fig.width=10, message=FALSE, paged.print=FALSE}
p2
```
```{r}
pathway_multi <- names(abundance_CPM)[1:13]

p_list <- list()
for(i in seq(length(pathway_multi))){
pathway <- pathway_multi[i]
p_list[[i]] <- plot_ordination(tse, paste("UMAP", add_name, sep=""), pathway=pathway, data_type="Functional")$p_1spec 
}
```


```{r}
ggsave(paste(path_analysis, "Plots/Ordination_plots/UMAP", add_name,".png", sep=""),
       p1 + p5 + p2+ p4 + p6 + p3+ plot_layout(guides = 'collect'), 
       width=30, height=15)

ggsave(paste(path_analysis, "Plots/Ordination_plots/UMAP", add_name,"_withAbundance.png", sep=""), 
       wrap_plots(p_list, ncol=4) , 
       width=30, height=20)
```

```{r}
df_list <- list(PartHsd_PCoA_BC, PartHsd_NMDS_BC,
                PartHsd_TSNE, PartHsd_UMAP)
PartHsd <- df_list %>% purrr::reduce(dplyr::full_join, by='Participant_ID')
PartHsd <- PartHsd[order(PartHsd$Participant_ID),]
PartHsd <- dplyr::left_join(PartHsd, as.data.frame(subset(colData(tse), Timepoint=="T2")[, c(1,4:6)]), by="Participant_ID")   


flextable(PartHsd)
```


```{r paged.print=FALSE}
PartHsd$TotalOut <- rowSums(!is.na(PartHsd[, 2:(length(df_list)+1)]))

flextable(subset(PartHsd, TotalOut>=3))
```
  

## Saving tse
```{r}
save(tse, tse_pathOnly, file=paste(path_analysis, "tse_allwithordination.Rdata", sep=""))
# data with ordination 
# To keep same ordination (ran on Computer office)
```


# Bar plots 
```{r fig.height=12, fig.width=12, message=FALSE, warning=FALSE}
n_taxa <- 50
cols_mod <- cols
cols_mod[n_taxa+1] <- "lightgrey"

physeq_noFact  %>%
  ps_select(Participant_ID, Inter_treat) %>% 
  phyloseq::merge_samples(group = "Inter_treat") %>%
  comp_barplot(tax_level = "Pathway", n_taxa = n_taxa, bar_width = 0.9, palette=cols_mod,
               tax_transform_for_plot="identity", bar_outline_colour = NA, 
               sample_order=c("Before.Yogurt", "After.Yogurt", "Before.YogurtOatmeal", "After.YogurtOatmeal")) + 
  labs(x = NULL, y = NULL) + guides(fill = guide_legend(ncol = 1, reverse=TRUE)) +theme(legend.key.size = unit(0.5, 'cm'))

physeq_pathOnly_noFact  %>%
  ps_select(Participant_ID, Inter_treat) %>% 
  phyloseq::merge_samples(group = "Inter_treat") %>%
  comp_barplot(tax_level = "Pathway", n_taxa = n_taxa, bar_width = 0.9, palette=cols_mod,
               tax_transform_for_plot="identity", bar_outline_colour = NA, 
               sample_order=c("Before.Yogurt", "After.Yogurt", "Before.YogurtOatmeal", "After.YogurtOatmeal")) + 
  labs(x = NULL, y = NULL) + guides(fill = guide_legend(ncol = 1, reverse=TRUE)) +theme(legend.key.size = unit(0.5, 'cm'))

```

```{r fig.height=20, fig.width=20, message=FALSE, warning=FALSE}
physeq %>%
  comp_barplot("Pathway", facet_by  = "Inter_treat", n_taxa = n_taxa, label = "Participant_ID", bar_outline_colour=NA, bar_width=0.9, tax_transform_for_plot="identity",  palette=cols_mod, ncol=4)+
  coord_flip() + guides(fill = guide_legend(ncol = 1))

physeq_pathOnly %>%
  comp_barplot("Pathway", facet_by  = "Inter_treat", n_taxa = n_taxa, label = "Participant_ID", bar_outline_colour=NA, bar_width=0.9, tax_transform_for_plot="identity",  palette=cols_mod, ncol=4)+
  coord_flip() + guides(fill = guide_legend(ncol = 1))
```

# Beta diversity

```{r message=FALSE}
metadata <- as.data.frame(colData(tse))
assay_t <- t(assays(tse)[["CPM"]]) %>% as.data.frame()
assay_t2 <- assay_t
pathways <- colnames(assay_t2)
assay_t2$Sample_ID <- rownames(assay_t2)
assay_t2 <- merge(metadata, assay_t2, by="Sample_ID")

assay_t_Y <- subset(assay_t2,  Intervention== "Yogurt")
assay_t_YO <- subset(assay_t2,  Intervention== "YogurtOatmeal")
assay_t_Y <- assay_t_Y[, pathways]
assay_t_YO <- assay_t_YO[, pathways]
```

```{r}
Perct0 <- apply(assay_t, 2, function(x) sum(x==0)/nrow(assay_t)*100)
Perct0_Y <- apply(assay_t_Y, 2, function(x) sum(x==0)/nrow(assay_t_Y)*100)
Perct0_YO <- apply(assay_t_YO, 2, function(x) sum(x==0)/nrow(assay_t_YO)*100)

hist(Perct0)
```



```{r}
# path_infos <- data.frame(Perct0)
# path_infos$Pathways <- rownames(path_infos)
# path_infos$prev <- 100-path_infos$Perct0
# path_infos <- path_infos[, c(2,1,3)]
```

```{r}
# path_perct0_0 <- names(Perct0[Perct0 <= 0])
# path_perct0_0_90 <- names(Perct0[Perct0 > 0 & Perct0 < 90])
# path_no_fitted <- path_infos[which(!(path_infos$Pathways %in% path_perct0_0) & !(path_infos$Pathways %in% path_perct0_0_90)), "Pathways"]
```

```{r}
# length(path_perct0_0)
# length(path_perct0_0_90)
# length(path_no_fitted)
# # [1] 220
# # [1] 273
# # [1] 113
```


```{r}
path_infos <- data.frame(Perct0, Perct0_Y, Perct0_YO)
path_infos$Pathways <- rownames(path_infos)
path_infos$prev <- 100-path_infos$Perct0
path_infos$prev_Y <- 100-path_infos$Perct0_Y
path_infos$prev_YO <- 100-path_infos$Perct0_YO
path_infos <- path_infos[, c(4,1:3,5:7)]
```

```{r}
# subset(path_infos, Perct0_YO==0 & Perct0>0) # 2   7
# subset(path_infos, Perct0_Y==0 & Perct0>0) # 27   7

path_perct0_0 <- subset(path_infos, Perct0 <= 0)$Pathways
path_perct0_0_90 <- subset(path_infos, Perct0 > 0 & Perct0 < 90)$Pathways
path_no_fitted <- path_infos[which(!(path_infos$Pathways %in% path_perct0_0) & !(path_infos$Pathways %in% path_perct0_0_90)), "Pathways"]

path_perct0_0_Y <- subset(path_infos, Perct0_Y <= 0)$Pathways
path_perct0_0_90_Y <- subset(path_infos, Perct0_Y > 0 & Perct0_Y < 90)$Pathways
path_no_fitted_Y <- path_infos[which(!(path_infos$Pathways %in% path_perct0_0_Y) & !(path_infos$Pathways %in% path_perct0_0_90_Y)), "Pathways"]

path_perct0_0_YO <- subset(path_infos, Perct0_YO <= 0)$Pathways
path_perct0_0_90_YO <- subset(path_infos, Perct0_YO > 0 & Perct0_YO < 90)$Pathways
path_no_fitted_YO <- path_infos[which(!(path_infos$Pathways %in% path_perct0_0_YO) & !(path_infos$Pathways %in% path_perct0_0_90_YO)), "Pathways"]
```

```{r}
length(path_perct0_0) # 220 Pathways have no zero values in both interventions
length(path_perct0_0_90) # 273 Pathways have zero values in at least one of the intervention
length(path_no_fitted) # 113 Pathways have more than 90% zero values over both interventions
# [1] 220
# [1] 273
# [1] 113

length(path_perct0_0_Y) # 247 (220+27) Pathways have no zero values in Yogurt
length(path_perct0_0_90_Y) # 244 Pathways have zero values in Yogurt
length(path_no_fitted_Y) # 115 Pathways have more than 90% zero values in Yogurt
# [1] 247
# [1] 244
# [1] 115

length(path_perct0_0_YO)  # 222 (220+2) Pathways have no zero values in YogurtOatmeal
length(path_perct0_0_90_YO)  # 271 Pathways have zero values in YogurtOatmeal
length(path_no_fitted_YO) # 113 Pathways have more than 90% zero values in YogurtOatmeal
# [1] 222
# [1] 271
# [1] 113
```

## transformation + covariates

**Old version: separation between patways with or without zeroes made on full subset.**

We use the same set of pathways as `path_to_use` for model full, Y and YO.
For pathways with no zeros only in Y (27 pathways) or YO (2 pathways) it will not be possible to run NBZIMM.
For the final model (without covariate) that we use in the paper (see next section) we have different set of `path_to_use` for each model to avoid this issue.


```{r}
# length(path_perct0_0) # 220
# length(path_perct0_0_90) # 273
# length(path_no_fitted) # 113
```



```{r message=FALSE}
assay_t <- t(assays(tse)[["CPM"]]) %>% as.data.frame()
pathways <- colnames(assay_t)
assay_t$Sample_ID <- rownames(assay_t)
metadata <- as.data.frame(colData(tse))
DF <- merge(metadata, assay_t, by="Sample_ID")
DF_subs <- prepare_DATA(DF, NULL)
```


```{r}
#### Model full
compar_var = "T"
DF_work=DF_subs$DF
compa = "full"
FE <- c("B", "T", "C1", "C2", "Geschlecht", "Alter")
RE <- "Participant_ID"

#### Model Y
compar_var = "B"
DF_work=DF_subs$DF_int1
compa = "Y"
FE <- c("B", "P", "Geschlecht", "Alter")
RE <- "Participant_ID"

#### Model YO
compar_var = "B"
DF_work=DF_subs$DF_int2
compa = "YO"
FE <- c("B", "P", "Geschlecht", "Alter")
RE <- "Participant_ID"

```

### LMER on pathways without zeroes
```{r}
path_to_use <- path_perct0_0
```


```{r}
model <- "lmer_bc"
trans <- "_cov"
sig_threshold <- 0.03
```

```{r}
trans_var <- rep("bc", length(path_to_use))
names(trans_var) <- path_to_use
DF_work_mod <- transform_data(DF_work, trans_var)
measures_Trans <- sapply(names(trans_var), function(x) if(trans_var[[x]] !="") paste(x, trans_var[[x]], sep="_") else x)
```


```{r}
tic()
# lmer
wd0 <- paste(path_analysis, "ResultsBetadiv/", sep="")
setwd(wd0)
run_taxa_model(measures_Trans, package="lme4", counts=F, FE=FE, RE=RE, DF=DF_work_mod, compar_var=compar_var, var_check=NULL, savingName=paste("lmer_bc_", compa, sep=""), sig_threshold=sig_threshold, check_mod=F, save_mod=F,  path=paste("lmer_bc_cov/", compa, "/", sep=""))
toc()
# Model full - long because of check_mod=T
# Model Y: 203.73 sec elapsed
# Model YO: 188.5 sec elapsed
```


```{r}
# Only check model for compa=full
bad_fit_LM <- c("ASPASN_PWY", "COLANSYN_PWY", "FASYN_ELONG_PWY", "P23_PWY", "PHOSLIPSYN_PWY", "PWY_5659", "PWY_6876", "PWY_6897", "PWY_7221", "PWY_7357", "PWY_7456", "PWY0_1586", "PWY66_429", "THISYNARA_PWY")
good_fit_LM <- path_perct0_0[!(path_perct0_0 %in% bad_fit_LM)]

path_infos$fit_LM <- case_when(
  path_infos$Pathways  %in% bad_fit_LM ~ "Bad fit",
  path_infos$Pathways  %in% good_fit_LM ~ "good fit",
  .default = "no fitted") 

ggplot(subset(path_infos, fit_LM != "no fitted"), aes(x=Perct0, fill=fit_LM)) + 
  geom_histogram(alpha=0.6, position = 'identity') +
  scale_fill_manual(values=c("#404080", "#69b3a2", "pink")) 
```
### CPLM on pathways with zeroes

```{r}
model <- "glmmTMB"
trans <- "_cov"
sig_threshold <- 0.03
```


```{r}
tic()
# CPLM
wd0 <- paste(path_analysis, "ResultsBetadiv/", sep="")
setwd(wd0)
run_taxa_model(path_perct0_0_90[-c(56, 57)], package="glmmTMB", counts=F, FE=FE, RE=RE, DF=DF_work, compar_var=compar_var, var_check=NULL, savingName=paste("glmmTMB_", compa, sep=""), sig_threshold=sig_threshold, check_mod=F, save_mod=F,  path=paste("glmmTMB_cov/", compa, "/", sep=""))
toc()
# Model full: 376.56 sec elapsed
# Model Y: 477.94 sec elapsed  R stop for PWY_5367, PWY_6531 (id 56, 57)
# Model YO: 350.66 sec elapsed
```

### ZIG on pathways with zeroes
```{r}
library(NBZIMM)
```


```{r}
model <- "NBZIMM_asinsqrt"
trans <- "_cov"
sig_threshold <- 0.03
```

```{r}
DF_work_mod=DF_work
DF_work_mod[,pathways] <- asin(sqrt(DF_work_mod[, pathways]/1000000)) 
```


```{r}
tic()
# ZIG
wd0 <- paste(path_analysis, "ResultsBetadiv/", sep="")
setwd(wd0)
run_taxa_model(path_perct0_0_90, package="NBZIMM", counts=F, FE=FE, RE=RE, DF=DF_work_mod, compar_var=compar_var, var_check=NULL, savingName=paste("NBZIMM_asinsqrt_", compa, sep=""), sig_threshold=sig_threshold, check_mod=F, save_mod=F,  path=paste("NBZIMM_asinsqrt_cov/", compa, "/", sep=""))
toc()
# Model full: 190.37 sec elapsed
# Model Y: 191.65 sec elapsed
# Model YO: 160.56 sec elapsed
```





## transformation no covariates

**New version: separation between patways with or without zeroes made on each subsets.**

Final model that we use in the paper, we have different set of `path_to_use` for each model to be sure to have results on all pathways.

```{r}
### saved in Rosner_Model_Trans
# same DF and measures_Trans as defined in previous section
```


```{r}
#### Model full
compar_var = "T"
DF_work=DF_subs$DF
compa = "full"
FE <- c("B", "T", "C1", "C2")
RE <- "Participant_ID"
path_to_use <- path_perct0_0
path_to_use_w0 <- path_perct0_0_90

#### Model Y
compar_var = "B"
DF_work=DF_subs$DF_int1
compa = "Y"
FE <- c("B", "P")
RE <- "Participant_ID"
path_to_use <- path_perct0_0_Y
path_to_use_w0 <- path_perct0_0_90_Y

#### Model YO
compar_var = "B"
DF_work=DF_subs$DF_int2
compa = "YO"
FE <- c("B", "P")
RE <- "Participant_ID"
path_to_use <- path_perct0_0_YO
path_to_use_w0 <- path_perct0_0_90_YO
```

### LMER on pathways without zeroes
```{r}
# path_to_use <- path_perct0_0
```

```{r}
model <- "lmer_bc"
trans <- ""
sig_threshold <- 0.03
```

```{r}
trans_var <- rep("bc", length(path_to_use))
names(trans_var) <- path_to_use
DF_work_mod <- transform_data(DF_work, trans_var)
measures_Trans <- sapply(names(trans_var), function(x) if(trans_var[[x]] !="") paste(x, trans_var[[x]], sep="_") else x)
```


```{r}
tic()
# lmer
wd0 <- paste(path_analysis, "ResultsBetadiv/", sep="")
setwd(wd0)
run_taxa_model(measures_Trans, package="lme4", counts=F, FE=FE, RE=RE, DF=DF_work_mod, compar_var=compar_var, var_check=NULL, savingName=paste("lmer_bc_", compa, sep=""), sig_threshold=sig_threshold, check_mod=F, save_mod=F,  path=paste("lmer_bc/", compa, "/", sep=""))
toc()
# Model full: 194.9 sec elapsed
# Model Y: 163.7 sec elapsed / 170.03 sec elapsed
# Model YO: 146.71 sec elapsed / 135.71 sec elapsed
```


### CPLM on pathways with zeroes

```{r}
model <- "glmmTMB"
trans <- ""
sig_threshold <- 0.03
```


Old version: use the same set of pathways as path_perct0_0_90 for model full, Y and YO. 
We didn't rerun CPLM for the final analysis.

```{r}
tic()
# CPLM
wd0 <- paste(path_analysis, "ResultsBetadiv/", sep="")
setwd(wd0)
run_taxa_model(path_perct0_0_90[-c(93, 94)], package="glmmTMB", counts=F, FE=FE, RE=RE, DF=DF_work, compar_var=compar_var, var_check=NULL, savingName=paste("glmmTMB_", compa, sep=""), sig_threshold=sig_threshold, check_mod=F, save_mod=F,  path=paste("glmmTMB/", compa, "/", sep=""))
toc()
# Model full: 350.92 sec elapsed
# Model Y: 410.14 sec elapsed  R stop for GALACTARDEG_PWY, GLUCARGALACTSUPER_PWY (id 93, 94)
# Model YO: 346.25 sec elapsed
```

### ZIG on pathways with zeroes

```{r}
model <- "NBZIMM_asinsqrt"
trans <- ""
sig_threshold <- 0.03
```

```{r}
DF_work_mod=DF_work
DF_work_mod[,pathways] <- asin(sqrt(DF_work_mod[, pathways]/1000000)) 
```

```{r}
tic()
# ZIG
wd0 <- paste(path_analysis, "ResultsBetadiv/", sep="")
setwd(wd0)
run_taxa_model(path_to_use_w0, package="NBZIMM", counts=F, FE=FE, RE=RE, DF=DF_work_mod, compar_var=compar_var, var_check=NULL, savingName=paste("NBZIMM_asinsqrt_", compa, sep=""), sig_threshold=sig_threshold, check_mod=F, save_mod=F,  path=paste("NBZIMM_asinsqrt/", compa, "/", sep=""))
toc()
# Model full: 166.75 sec elapsed
# Model Y: 143.54 sec elapsed / 133.41 sec elapsed
# Model YO: 136.61 sec elapsed / 126.8 sec elapsed
```




## Redo important plots
```{r}
#### need to specify the following parameters from models defined above:
# compar_var = 
# compa = 
# FE <- 
# RE <- 
# model <- 
# DF_work_mod <-
```


```{r}
### REDO all plots at new sig_threshold
## Careful to have the correct DF_work_mod!! 

sig_threshold <- 0.03

path <- paste(path_analysis, "ResultsBetadiv/", model, trans, "/", compa,"/", sep="")
res_all <- read.xlsx(file=paste(path, model, "_", compa, "_results.xlsx", sep="") , sheetName="res_all")
res_anov_all <- read.xlsx(file=paste(path, model, "_", compa, "_results.xlsx", sep="") , sheetName="res_anov_all")
PRS_all <- read.xlsx(file=paste(path, model, "_", compa, "_results.xlsx", sep="") , sheetName="PRS_all")

savingName=paste(model, "_", compa, sep="")
log_file <- file.path(path, "run_taxa_model.log")
logging::addHandler(logging::writeToFile, file = log_file, level = "DEBUG")
logging::logdebug("recreate plots with sig_threshold: %s", sig_threshold)

plot_results(res_all, res_anov_all, PRS_all, DF=DF_work_mod, FE, sig_threshold, savingName, path)
```


```{r}
### REDO Forest plot for better size
path <- paste(path_analysis, "ResultsBetadiv/", model, trans, "/", compa,"/", sep="")
res_all <- read.xlsx(file=paste(path, model, "_", compa, "_results.xlsx", sep="") , sheetName="res_all")
sig_res_all <- subset(res_all, p.value < sig_threshold) 


sig_res_all$color <- case_when(sig_res_all$q.value < 0.05 ~  "q.val<0.05",
                               sig_res_all$q.value >= 0.05 ~ "q.val>=0.05")
var <- unique(sig_res_all$names)
ll_plot <- list()
for(i in seq(var)){
  sub <- subset(sig_res_all, names==var[i])
  sub$feature <- factor(sub$feature, levels = rev(sub$feature))
  pp <- ggplot(sub, aes(x=est, y=feature)) +  geom_point(size=1.2, aes(color=color)) +
    scale_colour_manual(values= c(`q.val>=0.05`="black", `q.val<0.05`="#2196F3"))+
    geom_vline(xintercept = 0, linetype = "dashed", color = "grey40", linewidth=0.5) + 
    xlab("effect size") + ylab("") + ggtitle(var[i]) + theme_minimal()
  if(all(is.na(sub$ci_low))) ll_plot[[i]] <- pp
  else  ll_plot[[i]] <- pp + geom_errorbar(aes(x = est, xmin = ci_low, xmax = ci_high, 
                                               color=color), width = 0.2, linewidth=0.5) 
}

savingName <- paste(model, "_", compa, "_res_all_betterSize", sep="")
ggsave_adaptSize(paste(path, savingName, ".png", sep=""), ll_plot, unit_width=5, unit_height=6)
```



```{r}
### Forest plot on PRS results
path <- paste(path_analysis, "ResultsBetadiv/", model, trans, "/", compa,"/", sep="")
PRS_all <- read.xlsx(file=paste(path, model, "_", compa, "_results.xlsx", sep="") , sheetName="PRS_all")
sig_res_PRS <- subset(PRS_all, p.value < sig_threshold) 

sig_res_PRS$color <- case_when(sig_res_PRS$q.value < 0.05 ~  "q.val<0.05",
                               sig_res_PRS$q.value >= 0.05 ~ "q.val>=0.05")
sig_res_PRS$feature <- factor(sig_res_PRS$feature, levels = rev(sig_res_PRS$feature))

if(compar_var == "B") xlab <- "After - Baseline"
if(compar_var == "T") xlab <- "After_YogurtOatmeal - Other"

pp <- ggplot(sig_res_PRS, aes(x=estimate, y=feature)) + geom_point(size=1.2, aes(color=color)) +
    scale_colour_manual(values= c(`q.val>=0.05`="black", `q.val<0.05`="#2196F3"))+
    geom_vline(xintercept = 0, linetype = "dashed", color = "grey40", linewidth=0.5) + 
    xlab(xlab) + ylab("") + theme_minimal()
pp  <- pp + geom_errorbar(aes(x = estimate, xmin = lower.CL, xmax = upper.CL, 
                                               color=color), width = 0.2, linewidth=0.5) 

savingName <- paste(model, "_", compa, "_res_PRS", sep="")
ggsave(paste(path, savingName, ".png", sep=""), pp, width=6, height=6)
```


```{r}
# Print taxa plot for significant taxa in model full for better visualization
path <- paste(path_analysis, "ResultsBetadiv/", model, trans, "/full/", sep="")
PRS_all <- read.xlsx(file=paste(path, model, "_full_results.xlsx", sep="") , sheetName="PRS_all")
sig_res_PRS <- subset(PRS_all, p.value < sig_threshold) 

taxa_to_plot <- sig_res_PRS$feature

## Careful to have the correct DF_work_mod!! 
###### if LMER model check that DF_work_mod correspond to bc transfromation
###### if ZIG model:
# DF_work_mod[,taxa] <- asin(sqrt(DF_work_mod[, taxa]/100))


for(i in seq(length(taxa_to_plot))){
  taxa_to_plot_i <- taxa_to_plot[i]
  
  p_rel <- plot_taxa_variable(taxa_to_plot_i, DF_work_mod, counts=F)
  
  name_plot <- taxa_to_plot_i
  if(nchar(name_plot) > 28) name_plot <- substr(name_plot, 1, 28)
  
  ggsave(paste(path, "/Plots/Plot_PRStaxa/", i, "_", name_plot, ".png", sep=""),
         p_rel+ plot_annotation(title = taxa_to_plot_i), 
         width=10, height=10)
}
```






